### ISSUE TITLE
Misaligned implementation of top_k sampling with huggingface/transformers

### ISSUE THREAD

**Issue by Abraham-Xu** [vllm-project/vllm#718]
If we want to do a top_k sampling in vllm by setting top_k>0:
1. vllm computes softmax of logits firstly
2. then does top_k truncation afterwards to obtain the final probability distribution ( which is NOT normalized ).
3. then does sample by torch.multinomial operation.

If we want to do a top_k sampling in  huggingface/transformers by setting top_k>0 and do_sampling=True. 
1. transformers computes topk truncation with logits firstly.
2. then computes the softmax of the truncated logits ( which is normalized ).
3. then does sample by torch.multinomial operation.
This difference will result in different top_k sampling generated tokens even the same random seed was set in vllm and transformers. I wonder whether it is a bug of vllm or different styles of top_k sampling implementations.

vllm:
vllm/model_executor/layers/sampler.py
```
       # We use float32 for probabilities and log probabilities.
        # Compute the probabilities.
        probs = torch.softmax(logits, dim=-1, dtype=torch.float)
        # Compute the log probabilities (before applying top-p and top-k).
        logprobs = torch.log(probs)

        # Apply top-p and top-k truncation.
        top_ps, top_ks = _get_top_p_top_k(input_metadata, self.vocab_size)
        assert len(top_ps) == len(top_ks) == probs.shape[0]
        do_top_p = any(p < 1.0 - _SAMPLING_EPS for p in top_ps)
        do_top_k = any(k != self.vocab_size for k in top_ks)
        if do_top_p or do_top_k:
            probs = _apply_top_p_top_k(probs, top_ps, top_ks)
```
huggingface/transformers:
generation/utils.py
```
            # pre-process distribution
            next_token_scores = logits_processor(input_ids, next_token_logits)
            next_token_scores = logits_warper(input_ids, next_token_scores)

            # Store scores, attentions and hidden_states when required
            if return_dict_in_generate:
                if output_scores:
                    scores += (next_token_scores,)
                if output_attentions:
                    decoder_attentions += (
                        (outputs.decoder_attentions,) if self.config.is_encoder_decoder else (outputs.attentions,)
                    )
                    if self.config.is_encoder_decoder:
                        cross_attentions += (outputs.cross_attentions,)

                if output_hidden_states:
                    decoder_hidden_states += (
                        (outputs.decoder_hidden_states,)
                        if self.config.is_encoder_decoder
                        else (outputs.hidden_states,)
                    )

            # sample
            probs = nn.functional.softmax(next_token_scores, dim=-1)
            next_tokens = torch.multinomial(probs, num_samples=1).squeeze(1)
```
top_k truncation was done in logits_warper before softmax.
generation/logits_process.py
```
class TopKLogitsWarper(LogitsWarper):
    def __init__(self, top_k: int, filter_value: float = -float("Inf"), min_tokens_to_keep: int = 1):
        if not isinstance(top_k, int) or top_k <= 0:
            raise ValueError(f"`top_k` has to be a strictly positive integer, but is {top_k}")

        self.top_k = max(top_k, min_tokens_to_keep)
        self.filter_value = filter_value

    @add_start_docstrings(LOGITS_PROCESSOR_INPUTS_DOCSTRING)
    def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> torch.FloatTensor:
        top_k = min(self.top_k, scores.size(-1))  # Safety check
        # Remove all tokens with a probability less than the last token of the top-k
        indices_to_remove = scores < torch.topk(scores, top_k)[0][..., -1, None]
        scores = scores.masked_fill(indices_to_remove, self.filter_value)
        return scores
```

**Comment by lucasjinreal** [comment]
This might be the issue of unaligned result compare with hf and vllm interms of same generation params and inputs.
Does there any solution for this?

**Comment by kechunFIVE** [comment]
In vLLM 0.1.3, vLLM's top-k seems to be aligned with hf, but in my tests (baichuan-13B), the results are still a little worse than HF.

`#Apply top-p and top-k truncation.
        top_ps, top_ks = _get_top_p_top_k(input_metadata, self.vocab_size)
        assert len(top_ps) == len(top_ks) == logits.shape[0]
        do_top_p = any(p < 1.0 - _SAMPLING_EPS for p in top_ps)
        do_top_k = any(k != self.vocab_size for k in top_ks)
        if do_top_p or do_top_k:
            logits = _apply_top_p_top_k(logits, top_ps, top_ks)

        # We use float32 for probabilities and log probabilities.
        # Compute the probabilities.
        probs = torch.softmax(logits, dim=-1, dtype=torch.float)
        # Compute the log probabilities (before applying top-p and top-k).
        logprobs = torch.log(probs)`

**Comment by kechunFIVE** [comment]
In vLLM 0.1.3, the  top-k  implemention seems to be aligned with hf, but in my test (baichuan-13B), the results are still a little worse than HF.
```#Apply top-p and top-k truncation.
        top_ps, top_ks = _get_top_p_top_k(input_metadata, self.vocab_size)
        assert len(top_ps) == len(top_ks) == logits.shape[0]
        do_top_p = any(p < 1.0 - _SAMPLING_EPS for p in top_ps)
        do_top_k = any(k != self.vocab_size for k in top_ks)
        if do_top_p or do_top_k:
            logits = _apply_top_p_top_k(logits, top_ps, top_ks)

        # We use float32 for probabilities and log probabilities.
        # Compute the probabilities.
        probs = torch.softmax(logits, dim=-1, dtype=torch.float)
        # Compute the log probabilities (before applying top-p and top-k).
        logprobs = torch.log(probs)```

**Comment by lucasjinreal** [comment]
@kechunFIVE  what do u mean by a little? Does the sample forced? Can u post the params used in vllm and one in hf?

**Comment by hmellor** [comment]
Closing this issue as stale as there has been no discussion in the past 3 months.

If you are still experiencing the issue you describe, feel free to re-open this issue.
